src/django_components/dependencies.py

```diff
@@ -147,32 +152,114 @@ def _cache_script(
     comp_media_cache.set(cache_key, script.strip())
 
 
-def cache_inlined_js(comp_cls: Type["Component"], content: str) -> None:
-    if not _is_nonempty_str(comp_cls.js):
-        return
+def cache_component_js(comp_cls: Type["Component"]) -> None:
+    """
+    Cache the content from `Component.js`. This is the common JS that's shared
+    among all instances of the same component. So even if the component is rendered multiple
+    times, this JS is loaded only once.
+    """
+    if not comp_cls.js or not is_nonempty_str(comp_cls.js) or _is_script_in_cache(comp_cls, "js", None):
+        return None
+
+    _cache_script(
+        comp_cls=comp_cls,
+        script=comp_cls.js,
+        script_type="js",
+        input_hash=None,
+    )
+
+
+# NOTE: In CSS, we link the CSS vars to the component via a stylesheet that defines
+# the CSS vars under `[data-djc-css-a1b2c3]`. Because of this we define the variables
+# separately from the rest of the CSS definition.
+#
+# We use conceptually similar approach for JS, except in JS we have to manually associate
+# the JS variables ("stylesheet") with the target HTML element ("component").
+#
+# It involves 3 steps:
+# 1. Register the common logic (equivalent to registering common CSS).
+#    with `Components.manager.registerComponent`.
+# 2. Register the unique set of JS variables (equivalent to defining CSS vars)
+#    with `Components.manager.registerComponentData`.
+# 3. Actually run a component's JS instance with `Components.manager.callComponent`,
+#    specifying the components HTML elements with `component_id`, and JS vars with `input_hash`.
+def cache_component_js_vars(comp_cls: Type["Component"], js_vars: Dict) -> Optional[str]:
+    if not is_nonempty_str(comp_cls.js):
+        return None
+
+    # The hash for the file that holds the JS variables is derived from the variables themselves.
+    json_data = json.dumps(js_vars)
+    input_hash = md5(json_data.encode()).hexdigest()[0:6]
 
-    # Prepare the script that's common to all instances of the same component
-    # E.g. `my_table.js`
-    if not _is_script_in_cache(comp_cls, "js"):
+    # Generate and cache a JS script that contains the JS variables.
+    if not _is_script_in_cache(comp_cls, "js", input_hash):
         _cache_script(
             comp_cls=comp_cls,
-            script=content,
+            script="",  # TODO
```

Comment:
One thing that's missing from `cache_component_js_vars` and `cache_component_css_vars` is that we need to pre-process the JS / CSS vars into the correct shape so they could be imported.

E.g. for CSS, when we get a dictionary of values, we actually want to format it as proper CSS, so as
```css
[data-djc-css-a1b2c3] {
  --border_width: 3px;
  --button_color: blue;
}
```

I didn't include that in this PR for simplicity, hence why the `script="",  # TODO`.

This will be included in the PRs that actually add the JS / CSS vars features.